Sources/SwiftParser/Directives.swift

```diff
@@ -106,29 +109,56 @@ extension Parser {
         loopProgress.evaluate(self.currentToken)
       {
-        var (unexpectedBeforePoundIf, poundIf) = self.eat(poundIfHandle)
-        firstIteration = false
-        // Parse the condition.
+        var unexpectedBeforePound: RawUnexpectedNodesSyntax?
+        var pound: RawTokenSyntax
         let condition: RawExprSyntax?
-        switch poundIf.tokenKind {
-        case .poundIfKeyword, .poundElseifKeyword:
+        var atElifTypo: Bool {
+          guard self.at(TokenSpec(.pound)), self.currentToken.trailingTriviaText.isEmpty else {
+            return false
+          }
+          let identifierSpec = TokenSpec(.identifier, allowAtStartOfLine: false)
+          var lookahead = self.lookahead()
+          lookahead.consumeAnyToken()
+          guard lookahead.at(identifierSpec), lookahead.currentToken.tokenText == "elif", lookahead.currentToken.leadingTriviaText.isEmpty else {
+            return false
+          }
+          lookahead.consumeAnyToken()
+          return lookahead.at(identifierSpec)
+        }
+
+        if atElifTypo {
+          (unexpectedBeforePound, pound) = self.eat(poundIfHandle)
+          guard let identifier = self.consume(if: TokenSpec(.identifier, allowAtStartOfLine: false)) else {
+            preconditionFailure("The current token should be an identifier, guaranteed by the previous if statement")
+          }
+          unexpectedBeforePound = RawUnexpectedNodesSyntax(combining: unexpectedBeforePound, pound, identifier, arena: self.arena)
+          pound = self.missingToken(.poundElseifKeyword)
           condition = RawExprSyntax(self.parseSequenceExpression(.basic, forDirective: true))
-        case .poundElseKeyword:
-          if let ifToken = self.consume(if: .init(.if, allowAtStartOfLine: false)) {
-            unexpectedBeforePoundIf = RawUnexpectedNodesSyntax(combining: unexpectedBeforePoundIf, poundIf, ifToken, arena: self.arena)
-            poundIf = self.missingToken(.poundElseifKeyword)
+        } else {
+          (unexpectedBeforePound, pound) = self.eat(poundIfHandle)
+          firstIteration = false
+          switch pound.tokenKind {
+          case .poundIfKeyword, .poundElseifKeyword:
             condition = RawExprSyntax(self.parseSequenceExpression(.basic, forDirective: true))
-          } else {
+          case .poundElseKeyword:
+            if let ifToken = self.consume(if: .init(.if, allowAtStartOfLine: false)) {
+              unexpectedBeforePound = RawUnexpectedNodesSyntax(combining: unexpectedBeforePound, pound, ifToken, arena: self.arena)
+              pound = self.missingToken(.poundElseifKeyword)
+              condition = RawExprSyntax(self.parseSequenceExpression(.basic, forDirective: true))
+            } else {
+              condition = nil
+            }
+          case .pound:
             condition = nil
```

Comment:
@ahoppen I managed to fix the parser round-trip failure by setting `condition` to `nil` here. As I said before, I'm not sure if this is correct.

I've also added the reduced test case to ensure this parses without any round-trip failures or crashes.

Reply:
I suspect the fix here is not because you set `condition` to `nil` here but because you added the `case .pound` and thus you’re not hitting the `preconditionFailure` in the `default` case. Because AFAICT `condition` is always `nil` when we hit this statement anyway.

---

I still think that the right implementation here would be to `break` out of the outer loop here by adding a label to the loop and breaking that. You can see that if you add a substructure test for `testIfConfigExpr36` – I usually do that by setting a breakpoint in `assertParse` and execute `e print(tree.debugInitCall(includeTrivia: false)` in the debugger and then adding the relevant output (here it would be the `IfConfigDeclSyntax` to the `substructure` parameter of `assertParse` + fix the remaining compilation errors.

In that case, you’ll have a `IfConfigClauseSyntax` where `poundKeyword` is a standalone `poundToken` while we expect `poundKeyword` to be one of `#if`, `#else`, etc.

Reply:
This is still incorrect. If we have a standalone `#` instead of a `#elseif`, we will happily eat it as the `#elseif` token and continue parsing. You need to break out of the loop if you see a standalone `#` that is not a `#elif`.

I think it will break in the following example, which AFAICt will not emit an error even though it should (haven’t tested though)

```swift
#if true
# someCondition
#endif
```

---

Also, I think you can move your `atElifTypo` inside this switch case. That way it‘s more performant because we don’t need to check for `#elif` in valid Swift code.